addons/binding/org.openhab.binding.plclogo/src/main/java/org/openhab/binding/plclogo/PLCLogoBindingConstants.java

```diff
@@ -0,0 +1,160 @@
+
+    public static final @NonNull String BINDING_ID = "plclogo";
+
+    // List of all thing type UIDs
+    public static final @NonNull ThingTypeUID THING_TYPE_DEVICE = new ThingTypeUID(BINDING_ID, "device");
+    public static final @NonNull ThingTypeUID THING_TYPE_ANALOG = new ThingTypeUID(BINDING_ID, "analog");
+    public static final @NonNull ThingTypeUID THING_TYPE_MEMORY = new ThingTypeUID(BINDING_ID, "memory");
+    public static final @NonNull ThingTypeUID THING_TYPE_DIGITAL = new ThingTypeUID(BINDING_ID, "digital");
+    public static final @NonNull ThingTypeUID THING_TYPE_DATETIME = new ThingTypeUID(BINDING_ID, "datetime");
+    public static final @NonNull ThingTypeUID THING_TYPE_PULSE = new ThingTypeUID(BINDING_ID, "pulse");
+
+    // List of all channels
+    public static final @NonNull String STATE_CHANNEL = "state";
+    public static final @NonNull String OBSERVE_CHANNEL = "observed";
+    public static final @NonNull String VALUE_CHANNEL = "value";
+    public static final @NonNull String RTC_CHANNEL = "rtc";
+
+    // List of all channel properties
+    public static final @NonNull String BLOCK_PROPERTY = "block";
+
+    // List of all item types
+    public static final @NonNull String ANALOG_ITEM = "Number";
+    public static final @NonNull String DATE_TIME_ITEM = "DateTime";
+    public static final @NonNull String DIGITAL_INPUT_ITEM = "Contact";
+    public static final @NonNull String DIGITAL_OUTPUT_ITEM = "Switch";
+
+    // LOGO! family definitions
+    public static final @NonNull String LOGO_0BA7 = "0BA7";
+    public static final @NonNull String LOGO_0BA8 = "0BA8";
+
+    // LOGO! diagnostics memory
+    public static final @NonNull Integer LOGO_STATE = 984; // Diagnostics
+
+    private static final Map<Integer, String> LOGO_STATES_0BA7 = new ImmutableMap.Builder<Integer, String>()
+    // @formatter:off
+        /*
+         *   put(   "VB", 0);
+         * - Network access errors
+         * - Expansion module bus errors
+         * - SD card read/write errors
+         * - SD card write protection
+         *
+         * - Netzwerkzugriffsfehler
+         * - Erweiterungsmodul-Busfehler
+         * - Fehler beim Lesen oder Schreiben der SD-Karte
+         * - Schreibschutz der SD-Karte
+         */
+        .build();
+    // @formatter:on
+
+    private static final Map<Integer, String> LOGO_STATES_0BA8 = new ImmutableMap.Builder<Integer, String>()
+    // @formatter:off
+            /*
+            *   put(   "VB", 0);
+            * - Ethernet link errors
+            * - Expansion module changed
+            * - SD card read/write errors
+            */
+        .put(8, "SD card does not exist")
+            /*
+            * - SD card is full
+            *
+            * - Netzwerk Verbindungsfehler
+            * - Ausgetauschtes Erweiterungsmodul
+            * - Fehler beim Lesen oder Schreiben der SD-Karte
+            *
+            * put(8, "SD-Karte nicht vorhanden");

```

Comment:
Have you an idea how to use openHAB translation service or where can i find some example?
I read documentation, but have still no clue, how to use it from code and not xml files.

Reply:
@cweitkamp & @lolodomo are our translation guru's :-)

Reply:
A question, if I may. What is the purpose of those translations? I looked through the code and could not find a line where the Map `LOGO_STATES` is used.

According to the documentation the binding can subscribe to the `TranslationProvider` and the `LocaleProvider` OSGi services to localize text. But the problem is that we have to use the `bundleContext` to retrieve a translation. The `bundleContext` is a deprecated property of the `BaseThingHandler` which must not be used by bindings.

```java
String text = i18nProvider.getText(bundleContext.getBundle(), "my.key", "DefaultValue", Locale.GERMAN);
```

Reply:
@cweitkamp : Thanks for reply. Sure, you may :-) There is no official documentation for error states.  It was a very hard job to get messages produced by different Logo families. Reading 1 byte at address 984 (LOGO_STATE) returns a bitmask, each bit maps to one error state. I was able to find meaning of 8 now, and sadly, have no idea about other bits. Maybe i or someone else get more clue next time. Additionally, as far i have seen, openHAB can not visualize string lists now. Anyway, for me, it's good point to start and my idea was to make as much as possible. Nevertheless, i will leave the implementation as is, since "properly" one is not possible :-(